casper/src/main/scala/coop/rchain/casper/util/ProtoUtil.scala

```diff
@@ -422,4 +426,77 @@ object ProtoUtil {
     (missingParents union missingJustifications).toList
   }
 
+  // Return hashes of all blocks that are yet to be seen by the passed in block
+  def unseenBlockHashes[F[_]: Sync: BlockStore](
+      dag: BlockDagRepresentation[F],
+      block: BlockMessage
+  ): F[Set[BlockHash]] =
+    for {
+      latestMessages        <- dag.latestMessages
+      latestMessagesOfBlock <- ProtoUtil.toLatestMessage(block.justifications, dag)
+      result <- latestMessages.toList
+                 .traverse {
+                   case (validator, latestMessage) =>
+                     latestMessagesOfBlock.get(validator) match {
+                       case Some(latestMessageOfBlockByValidator) =>
+                         DagOperations
+                           .bfTraverseF(List(latestMessage))(
+                             block =>
+                               getCreatorJustificationUnlessGoal(
+                                 dag,
+                                 block,
+                                 latestMessageOfBlockByValidator
+                               )
+                           )
+                           .map(_.blockHash)
+                           .toSet
+                       case None =>
+                         Set.empty[BlockHash].pure[F]
+                     }
+                 }
+                 .map(_.flatten.toSet)
+    } yield result -- latestMessagesOfBlock.values.map(_.blockHash) - block.blockHash
+
+  private def getCreatorJustificationUnlessGoal[F[_]: Sync: BlockStore](
+      dag: BlockDagRepresentation[F],
+      block: BlockMetadata,
+      goal: BlockMetadata
+  ): F[List[BlockMetadata]] =
+    ProtoUtil.creatorJustification(block) match {
+      case Some(Justification(_, hash)) =>
+        dag.lookup(hash).flatMap {
+          case Some(creatorJustification) =>
+            if (creatorJustification == goal) {
+              List.empty[BlockMetadata].pure[F]
+            } else {
+              List(creatorJustification).pure[F]
+            }
+          case None =>
+            Sync[F].raiseError[List[BlockMetadata]](
+              new RuntimeException(s"Missing block hash $hash in block dag.")
```

Comment:
when would a block hash be missing from the dag?

Reply:
It should never be (unless you directly fiddled with the filesystem) and thus the raiseError